book-cw/part-2-coding/regex.adoc

```diff
@@ -1,3 +1,348 @@
+----
+
+This probably looks both sloppy and intimidating.
+I still find these big patterns a bit challenging to read, let alone write, but regex isn't that hard to get comfortable with.
+I can't say I'm not enjoying the shock value, though.
+
+We are attempting to capture four groups wrapped in `( )` marks.
+Meanwhile, we're ignoring a fifth group (the one with the `Defaults to` string in it), as we have no use for that text.
+
+Once you've captured your groups, writing the parsing template is relatively straightforward.
+Simply use the `$n` token, where `n` is the capture order slot for that variable.
+This way they can be reproduced out of order.
+
+[source,regex]
+.Regex replace template for find/replace operation
+----
+$1::\n$2\n+\n[horizontal]\nDefault::: `$4`\n+\nRequired::: $3\n
+----
+
+The `\n` token denotes a newline marker, of course you see four numbered group tokens we discussed a moment ago.
+The rest of the markup you'll recognize from our AsciiDoc source example above.
+Let's look at what this find/replace operation produced -- you'll recommend most of the other replacement template in here.
+
+.Regex find and replace result
+----
+Volume::
+Modulates how loudly the music will play
++
+[horizontal]
+Default::: `5`
++
+Required::: Optional.
+
+
+Bass::
+Modulates the deeper frequencies of sound
++
+[horizontal]
+Default::: `7`
++
+Required::: Optional.
+
+
+Treble::
+Modulates the higher frequencies of sound
++
+[horizontal]
+Default::: `7`
++
+Required::: Optional.
+
+
+Booster::
+Hypes the whole system up
++
+[horizontal]
+Default::: ``
++
+Required::: Required.
+----
+
+This is not perfect.
+We'll still need to do another round or two of changes against this result to get it how we need it to look.
+But you'll see now why I captured the `.` character in `Optional.` and `Required.`.
+Our next move will be to run a simple (non-regex) find and replace procedure on the string `Required.`, which will miss the `Required` in `Required:::`.
+We want to replace `Required.` with `*Yes*`.
+
+Finally, we can replace `Optional.` with `No`.
+Now our document matches our target format precisely.
+
+[TIP]
+Truthfully, this entire transformation could have been performed with one replace statement using conditionals, but that is way beyond the scope of this primer.
+
+What did we save ourselves?
+This obviously depends on how hard it was to create the patterns.
+As soon as you get them right, you're pretty much done.
+And we did not have to manually remove quotation marks, insert carriage returns, numerous newlines, that repetitive `+` symbol, and all those infernal colons.
+If this example were about three times the size, I would certainly have opted for regex to save time.
+
+As a bonus, we got to solve a cool little puzzle!
+Seriously what more could a technical documentarian ask for?
+
+Proficient use of regex does not require memorization of all regex patterns -- they can always be looked up.
```

Comment:
This is a really good point. You might want to point to your favorite regex resources here. No need to reproduce.

Also, some regex testers, like regexr.com and regex101.com can be really helpful.